### ISSUE TITLE
Switch to predefined keyboard on input focus

### ISSUE THREAD

**Issue by gamesfrager** [sds100/KeyMapper#686]
Good day.
Let me start by saying this app is amazing.
The added helper of screenshot for tapping is unbelievably helpful. It shows that you care about UX.
I'm a webdeveloper and I can understand how complex this app is.

Now for my setup and I hope I'm not posting a duplicate here (I did search and didn't find similar one):

I have Android 9 (Pie) and I'm using a controller to navigate the interface (custom car setup).
Since I'm not using a keyboard, I have installed a custom keyboard https://play.google.com/store/apps/details?id=com.rederik.idrivekeyboard

Which acts on up and down d-pad keys to navigate and d-pad centre for selecting, you get the idea.
Now the issue is, when I'm on an input field (Google maps search field as an example) the iDrive keyboard shows up, and Key Mapper is still intercepting the keys.

Would it be possible to:
- in Key Mapper settings set a default keyboard from the available system keyboards
- switch to the predefined keyboard when focus is on an input field (or maybe detect if the predefined keyboard is on screen)
- switch back to Key Mapper when the predefined keyboard is off screen (dismissed)

This would really help me finish my car project.
Much appreciate your amazing work on this app. 

**Developer TODO (don't remove)**
- [ ] create new branch. put issue number at start of name
- [ ] write tests before working on feature. put issue number in comment
- [ ] update documentation
- [ ] squash and merge
- [ ] delete branch

**Comment by sds100** [comment]
Hello,
Very interesting use case for Key Mapper! This would only be possible if you can use ADB or root with your Android set-up. Key Mapper needs WRITE_SECURE_SETTINGS permission to be able to change the active keyboard without user-input. If you are able to do this then I will see if it is possible to detect when an input field is selected

**Comment by gamesfrager** [comment]
Thanks for the reply.
I spent all of last night and came to same conclusion :) I made it work with Tasker and AutoInput plugin since AutoInput can detect input focus and input focus out and basically using AutoTool's secure settings to disable the accessibility of Key Mapper.

As you can see, so many hoops to jump through and many possible fail points (Tasker crash, AutoInput crash etc.) and not to mention it takes ~1.5 seconds for all of this to happen. So it would be really nice to have it all native in key Mapper.
And I do have a rooted device.

Again, appreciate the reply.

**Comment by sds100** [comment]
Cool, I will have to find a way to reliably detect input focus.

**Comment by sds100** [comment]
[keymapper-2.4.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/6676840/keymapper-2.4.0-alpha-debug.zip)
@gamesfrager can you install this apk. This is a proof of concept just to see if the way I am detecting whether there is input focus works well. There is no UI for this feature yet so this apk won't trigger your key maps if there is input focus.

**Comment by gamesfrager** [comment]
Installed, enabled accessibility service and looking at the log when I focus on an input field I see:

```Received event from accessibility service: Pong(key=ping_service)```
I hope this is what you were looking for.

**Comment by sds100** [comment]
There isn't any log output for the feature. That is something else. I just want you to test it because it is implemented but I haven't added a UI to turn it on and off yet (the longest to implement bit).

**Comment by gamesfrager** [comment]
What would you like me to test? should I create some test maps and see if they trigger? or is there something specific you like me to test?

**Comment by sds100** [comment]
Yeah, I should be more clear. I'll make the apk automatically switch the keyboard like you requested - I forgot about that. Then you can test the feature by seeing if your remapped DPAD keys are temporarily disabled when you are trying to type something

**Comment by sds100** [comment]
[keymapper-2.4.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/6686403/keymapper-2.4.0-alpha-debug.zip)
This apk will automatically change the keyboard. Ensure that you have granted Key Mapper WRITE_SECURE_SETTINGS permission with ADB or root.

**Comment by gamesfrager** [comment]
Sorry it took a while to test.
I just tried it with a simple map (dpad-centre) to toggle play/pause of music.
The app has root permission, and accessibility enabled.
When I pres the button, the music playback is toggled.
When I'm on an input field, the iDrive keyboard does show up but pressing the dpad-centre still triggers the map (music playback is toggled) and the expected result is the selected letter from the iDrive keyboard is entered into the input field.

Is there a step(s) I'm missing? 

Here is the log https://drive.google.com/file/d/1Am5QQXvkqIbNdYyiWKRqIKa2lw_fgToF/view?usp=sharing

**Comment by sds100** [comment]
[keymapper-toggle-ime-input-focus-fix.zip](https://github.com/sds100/KeyMapper/files/6701295/keymapper-toggle-ime-input-focus-fix.zip)
maybe i sent the wrong apk because I don't see any of my logging for when input focus is lost/gained

**Comment by gamesfrager** [comment]
That could be it, since when I extracted the zip file, it asked if I want to replace the existing one or not. Will give this a shot tonight.

**Comment by gamesfrager** [comment]
Sorry again for taking so long to test this, been busy with family.
I just tried it, it still didn't stop capturing the iDrive controller button when on an input field and the iDrive soft keyboard is on screen.  

Here is the log file https://drive.google.com/file/d/1iSJpSrCxr8SyK98YbcdpU61rK_1cpKhE/view?usp=sharing

**Comment by sds100** [comment]
[keymapper-2.4.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/6732957/keymapper-2.4.0-alpha-debug.zip)
I have been really stupid. None of the apks were actually working 😂

**Comment by gamesfrager** [comment]
Success!! You got me puzzled a bit, as you named it similar to an earlier version you sent :D
But it worked 👍 and it worked great too. So thank you so much for this.

**Comment by sds100** [comment]
So everything works perfectly?

**Comment by gamesfrager** [comment]
Yes, everything seems to be working good. Thanks again.

**Comment by sds100** [comment]
implemented

**Comment by gamesfrager** [comment]
Hello. It's me again. This feature seems to have stopped when I installed the latest stable release. The beta that you sent before worked but the stable release doesn't work. The latest stable keymapper still intercepts the button presses and prevents the onscreen keyboard from working. Root access is granted everything is setup correctly.

**Comment by gamesfrager** [comment]
I have just confirmed with the alpha version that  it works properly. So I guess I'll keep using the alpha version until this is fixed on the stable release.
